x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java

```diff
@@ -93,34 +105,126 @@ protected void readInternal(byte[] b, int offset, int length) throws IOException
         }
     }
 
-    private void readInternalBytes(final long part, final long pos, byte[] b, int offset, int length) throws IOException {
+    private void readInternalBytes(final int part, long pos, final byte[] b, int offset, int length) throws IOException {
+        final long currentSequentialReadSize = sequentialReadSize;
+        if (currentSequentialReadSize != NO_SEQUENTIAL_READ_OPTIMIZATION) {
+            final StreamForSequentialReads streamForSequentialReads = streamForSequentialReadsRef.get();
+            if (streamForSequentialReads == null) {
+                // start a new sequential read
+                if (tryReadAndKeepStreamOpen(part, pos, b, offset, length, currentSequentialReadSize)) {
+                    return;
+                }
+            } else if (streamForSequentialReads.part == part && streamForSequentialReads.pos == pos) {
+                // continuing a sequential read that we started previously
+                assert streamForSequentialReads.isFullyRead() == false;
+                int read = streamForSequentialReads.inputStream.read(b, offset, length);
+                assert read <= length : read + " vs " + length;
+                streamForSequentialReads.pos += read;
+                position += read;
+                pos += read;
+                offset += read;
+                length -= read;
+
+                if (streamForSequentialReads.isFullyRead()) {
+                    if (streamForSequentialReadsRef.compareAndSet(streamForSequentialReads, null) != false) {
```

Comment:
just leave off the `!= false`

Reply:
33a1362ce45b035158f1243a5935548c8a5e782a